### ISSUE TITLE
Could not print Config Description for value xxx

### ISSUE THREAD

**Issue by knowthelist** [aaronjbrown/PyGrowatt#1]
```
~/src/PyGrowatt/scripts $ python growatt_wifi.py
2021-03-20 09:26:07,155 ERROR    Growatt:execute(499) Could not print Config Description for value 0x01
2021-03-20 09:26:08,257 ERROR    Growatt:execute(499) Could not print Config Description for value 0x02
2021-03-20 09:26:09,359 ERROR    Growatt:execute(499) Could not print Config Description for value 0x03
2021-03-20 09:26:10,458 INFO     Growatt:execute(497) Set Update Interval (0x04): 5
2021-03-20 09:26:11,564 INFO     Growatt:execute(497) Set Modbus Range low (0x05): 1
2021-03-20 09:26:12,658 INFO     Growatt:execute(497) Set Modbus Range high (0x06): 32
2021-03-20 09:26:13,758 INFO     Growatt:execute(497) Set UNKNOWN (0x07): X
2021-03-20 09:26:14,858 INFO     Growatt:execute(497) Set Device Serial Number (0x08): #########
2021-03-20 09:26:15,961 INFO     Growatt:execute(497) Set Hardware Version (0x09): XXXX
2021-03-20 09:26:17,055 INFO     Growatt:execute(497) Set UNKNOWN (0x0a): 0
2021-03-20 09:26:18,156 INFO     Growatt:execute(497) Set FTP credentials (0x0b): ##192.168.3.35/app/xml/#8081#
2021-03-20 09:26:19,258 INFO     Growatt:execute(497) Set DNS (0x0c): XXXXXXXXXXXXXXX
2021-03-20 09:26:20,358 INFO     Growatt:execute(497) Set UNKNOWN (0x0d): 16
2021-03-20 09:26:21,460 INFO     Growatt:execute(497) Set Local IP (0x0e): ########
2021-03-20 09:26:22,559 INFO     Growatt:execute(497) Set Local Port (0x0f): 80
2021-03-20 09:26:23,666 INFO     Growatt:execute(497) Set Mac Address (0x10): ########
2021-03-20 09:26:24,768 INFO     Growatt:execute(497) Set Server IP (0x11): #######
2021-03-20 09:26:24,774 ERROR    Growatt:execute(519) Could not find key
2021-03-20 09:26:25,865 INFO     Growatt:execute(497) Set Server Port (0x12): 5279
2021-03-20 09:26:26,968 INFO     Growatt:execute(497) Set Server (0x13): 
2021-03-20 09:26:28,065 INFO     Growatt:execute(497) Set Device Type (0x14): XXXXXXXXXXXXXXXXXXXX
2021-03-20 09:26:29,165 INFO     Growatt:execute(497) Set Software Version (0x15): 1.7.7.7
2021-03-20 09:26:30,267 INFO     Growatt:execute(497) Set Hardware Version (0x16): 1.7.6.0
2021-03-20 09:26:31,367 ERROR    Growatt:execute(499) Could not print Config Description for value 0x17
2021-03-20 09:26:32,468 ERROR    Growatt:execute(499) Could not print Config Description for value 0x18
2021-03-20 09:26:33,570 INFO     Growatt:execute(497) Set Netmask (0x19): 255.255.255.0
2021-03-20 09:26:34,667 ERROR    Growatt:execute(499) Could not print Config Description for value 0x1a
2021-03-20 09:26:35,766 ERROR    Growatt:execute(499) Could not print Config Description for value 0x1b
2021-03-20 09:26:36,867 ERROR    Growatt:execute(499) Could not print Config Description for value 0x1c
2021-03-20 09:26:37,964 ERROR    Growatt:execute(499) Could not print Config Description for value 0x1d
2021-03-20 09:26:39,065 INFO     Growatt:execute(497) Set Timezone (0x1e): XXXXXXXXXX
2021-03-20 09:26:40,174 INFO     Growatt:execute(497) Set Date (0x1f): 2021-03-20 15:55:01
2021-03-20 09:40:08,076 ERROR    Growatt:execute(499) Could not print Config Description for value 0x4c
2021-03-20 09:40:11,667 ERROR    sync:handle(212) Socket exception occurred Traceback (most recent call last):
  File "/home/mario/.local/lib/python2.7/site-packages/pymodbus/server/sync.py", line 202, in handle
    single=single)
  File "/home/mario/.local/lib/python2.7/site-packages/pymodbus/framer/socket_framer.py", line 153, in processIncomingPacket
    self._process(callback)
  File "/home/mario/.local/lib/python2.7/site-packages/PyGrowatt/growatt_framer.py", line 70, in _process
    callback(result)  # defer or push to a thread?
  File "/home/mario/.local/lib/python2.7/site-packages/pymodbus/server/sync.py", line 84, in execute
    response.transaction_id = request.transaction_id
AttributeError: 'NoneType' object has no attribute 'transaction_id'
```

Info: I replaced some of the confidentials with ##########
Some of the config keys are unknown (0x01, 0x02, 0x03, ... 0x4c). Can you add the missing ones?
Are these unknown keys the reason, that I can not receive PV data like Eac_today, Ppv ...

**Comment by knowthelist** [comment]
`Growatt:execute(519) Could not find key`
This was caused from missing [Growatt] ServerIP= in config.ini

These are my values of the UNKNOWN keys:
key=0x01  value=0
key=0x02  value=6.0
key=0x03  value=000
key=0x17  value=X
key=0x18  value=X
key=0x1a  value=#.#.#.#  -> seems to be the gateway ip
key=0x1b  value=1
key=0x1c  value=0
key=0x1d  value=X
key=0x4c  value=-86 -> seems to be the WiFi signal strength in dBm

**Comment by aaronjbrown** [comment]
Hello! Thanks for using the software and more importantly for providing the feedback!

Those keys were not mapped because they are not documented in the Growatt Protocol documents that I was able to find circulating on the internet, nor was I able to make sense of them in my limited personal setup. Thank you for providing your key/value pairs, they'll help narrow down the meaning of the "Query".

I see my gateway/router IP for 0x1a also, thanks for pointing this out! It also follows the netmask which is 0x19, so that would make sense. I'm not sure how it is configured/set, looking back through my recordings I often see it with what seems to be a default value of "192.168.5.1", but this may have been prior to me setting the inverter settings defined in the config file (Growatt.py:507-517).

How did you determine 0x4c was likely WiFi signal strength? It correlates with what I'm seeing (values from -74 to -77) but I haven't considered/tested this before. I could use an app on my phone standing in the location of my inverter to compare signal strength but it's night now so I cannot check until morning.

The values you see for 0x01-0x03, 0x17, 0x18, and 0x1b-0x1d are the same as I see; I can't infer anything of value from them :(

I see you've opened a Feature Request for MQTT support, does this mean you don't have a PVOutput account configured?

**Comment by knowthelist** [comment]
I have now looked a little deeper and think that the completeness of the config message is not my problem. I haven't received a single data message (0x0104) yet and I don't yet know the cause. First I want to see a simple EacToday parameter value in my console.

**Comment by aaronjbrown** [comment]
You're right, it looks like the ```2021-03-20 09:26:24,774 ERROR    Growatt:execute(519) Could not find key``` is because you don't have a ```ServerIP``` set in your config.ini but it won't stop you receiving the data. I assume you've manually configured your Wifi-X with your network information, and that's correctly displayed in the values you hashed out?

Without modifying the script it will only print EacToday after it is successfully uploaded to PVOutput. The quickest way you could test is replacing the pv_status_upload function (lines 40-86) with:
```
def pv_status_upload(datastore, interval):
    log.info("Eac_today: {}Wh)".format(datastore.getValues(4, inputRegisters["Eac_today"], 1)[0] * 100)) 

    # Keep repeating
    timer = threading.Timer(interval, pv_status_upload, args=(datastore, interval))
    timer.start()

    return
```
This should print Eac_today every 5 mins (as set in your Wifi-X ```2021-03-20 09:26:10,458 INFO     Growatt:execute(497) Set Update Interval (0x04): 5```). If the values are all 0, it means you're not receiving data.

There is also a scripts/Growatt.lua file, which is a Wireshark dissector. You can copy this into your Wireshark plugins directory and it will allow easier debugging of the packets between the Wifi-X and your computer. You can use a filter of ```growatt.fc == 0x04``` or ```growatt.fc == Energy``` to show only EnergyRequest and EnergyResponse PDUs. You should see the inverter send the data, then your server respond with an "ACK".

**Comment by aaronjbrown** [comment]
Hi @knowthelist, the bug causing the inverter to not send Energy data was fixed and merged to master in #4 along with clarifying the console prints of unknown Config Descriptions values and adding MQTT support.
